### PR TITLE
Fix PackageFeedSortingTest functional test

### LINKED ISSUE #6168: PackageFeedSortingTest functional test fails when < 10 packages in feed
The call to `responseText.IndexOf` does not account for the string not being found, due to reaching the end of the feed because the same package appears many times.

see: https://github.com/NuGet/NuGetGallery/blob/master/tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs#L211

From https://nuget.visualstudio.com/NuGetBuild/_build?buildId=34967
```
2018-07-16T16:25:48.4421714Z       Expected string e2e.semver2prerel.180630.045128.5737195 to come before 63712.  Expected list is: webgrease, owin, xsockets.plugin.framework, basetestpackage, microbuild.core, test20180627163712, newtonsoft.json, e2e.semver2prerel.180629.000605.2522011, e2e.semver2prerel.180630.045128.5737195, 63712
2018-07-16T16:25:48.4421714Z       Expected: True
2018-07-16T16:25:48.4421714Z       Actual:   False
2018-07-16T16:25:48.4421714Z       Stack Trace:
2018-07-16T16:25:48.4421714Z         C:\vsts-agent\_work\51\s\tests\NuGetGallery.FunctionalTests\ODataFeeds\V2FeedExtendedTests.cs(233,0): at NuGetGallery.FunctionalTests.ODataFeeds.V2FeedExtendedTests.<PackageFeedSortingTest>d__11.MoveNext()
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.4421714Z         --- End of stack trace from previous location where exception was thrown ---
2018-07-16T16:25:48.4421714Z            at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2018-07-16T16:25:48.4421714Z            at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2018-07-16T16:25:48.5047075Z ##[error]    NuGetGallery.FunctionalTests.ODataFeeds.V2FeedExtendedTests.PackageFeedSortingTest [FAIL]
```

### PR DESCRIPTION
**Pull request by chenriksson**
Fixes #6168 (functional test failure when <10 distinct packages in feed)

### CODE REVIEW

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs [hunk 1]
```diff
@@ -196,47 +197,67 @@ private static string GetPackagesAppearInFeedInOrderUrl(DateTime time, string ti
             var request = WebRequest.Create(UrlHelper.V2FeedRootUrl + @"stats/downloads/last6weeks/");
             var response = await request.GetResponseAsync();
 
-            string responseText;
-            using (var sr = new StreamReader(response.GetResponseStream()))
+            string responseText = await new StreamReader(response.GetResponseStream())
```
Comment (loic-sharma) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: Why was the using block for the `StreamReader` removed?

Comment (chenriksson) [reply tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: Oversight in refactoring - fixed!

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs [hunk 2]
```diff
@@ -193,50 +194,74 @@ private static string GetPackagesAppearInFeedInOrderUrl(DateTime time, string ti
...
-                    }
-                }
+                return await sr.ReadToEndAsync();
             }
+        }
 
-            request = WebRequest.Create(UrlHelper.BaseUrl + @"stats/packageversions");
+        private string[] GetPackageNamesFromFeedResponse(string feedResponseText)
+        {
+            const string PackageIdStartKey = @"""PackageId"": """;
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: IMO the `\"` syntax is a little easier to read, you've got like a billion `"`s in here

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs [hunk 3]
```diff
@@ -193,50 +194,74 @@ private static string GetPackagesAppearInFeedInOrderUrl(DateTime time, string ti
...
-                // We add angle brackets to prevent false failures due to duplicate package names in the page.
-                var condition = responseText.IndexOf(">" + packageName[i - 1] + "<", StringComparison.Ordinal) < responseText.IndexOf(">" + packageName[i] + "<", StringComparison.Ordinal);
-                Assert.True(condition, "Expected string " + packageName[i - 1] + " to come before " + packageName[i] + ".  Expected list is: " + packageName[0] + ", " + packageName[1] + ", " + packageName[2] + ", " + packageName[3] + ", " + packageName[4] + ", " + packageName[5] + ", " + packageName[6] + ", " + packageName[7] + ", " + packageName[8] + ", " + packageName[9]);
+                var start = seekStart(feedResponseText);
+                if (start < 0)
+                {
+                    break;
+                }
+                
+                feedResponseText = feedResponseText.Substring(start + PackageIdStartKey.Length);
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: The response is simple JSON, I think we could easily parse it with `JSON.NET` and this code would be much simpler.

```
var packages = JArray.Parse(feedResponseText);
foreach (var package in packages)
{
    yield return ((JObject)package)["PackageId"].Value<string>();
}
```

Could also model it and use `JsonConvert.DeserializeObject`, e.g.
```
private class Last6WeeksStatistics
{
    public IEnumerable<PackageStatistic> PackageStatistics { get; set; }
}

private class PackageStatistic
{
    public string PackageId { get; set; }
}
```

File: tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs [hunk 4]
```diff
@@ -193,50 +194,74 @@ private static string GetPackagesAppearInFeedInOrderUrl(DateTime time, string ti
...
+            var statsResponse = await GetResponseTextAsync(UrlHelper.BaseUrl + @"stats/packageversions");
+
+            var expectedNames = String.Join(", ", topDownloadsNames);
+            var last = topDownloadsNames.First();
+            foreach (var current in topDownloadsNames.Skip(1))
             {
-                responseText = await sr.ReadToEndAsync();
+                // Check to make sure the top 10 packages are in the same order as the feed.
+                // We add angle brackets to prevent false failures due to duplicate package names in the page.
+                var condition = statsResponse.IndexOf(">" + last + "<", StringComparison.Ordinal)
```
Comment (scottbommarito) [inline tests/NuGetGallery.FunctionalTests/ODataFeeds/V2FeedExtendedTests.cs]: This is pretty hacky--we're asserting the order of an HTML table for looking for the index of `>{package ID}<` and so on.

The real structure is:
```
<table>
    <tbody>
        <tr>
            <td>{row number}<td>
            <td><a href="/packages/{package ID}">{package ID}</a><td>
            ...
        </tr>
    </tbody>
</table>
```

I guess this is sufficient for now, but this code would be much more readable if you could find some sort of actual HTML parser for this purpose.

**Pull request merged**